miden-crypto/src/merkle/smt/partial.rs

```diff
@@ -813,4 +937,138 @@ mod tests {
+    /// - Key 1 has a value and a proof in the partial SMT.
+    /// - Key 3 has a value but is missing from the partial SMT (making node b non-empty).
+    /// - Keys 0, 2, 4, 5, 6, 7 are empty.
+    ///
+    /// Expected:
+    /// - Key 1: CAN update (explicitly tracked via proof)
+    /// - Key 0: CAN update (sibling of key 1, provably empty)
+    /// - Keys 4, 5, 6, 7: CAN update (in empty subtree f, provably empty)
+    /// - Keys 2, 3: CANNOT update (under non-empty node b, only have its hash)
+    #[test]
+    fn partial_smt_tracking_visualization() {
+        // Situation in the diagram mapped to depth-64 SMT.
+        const LEAF_0: u64 = 0;
+        const LEAF_1: u64 = 1 << 61;
+        const LEAF_2: u64 = 1 << 62;
+        const LEAF_3: u64 = (1 << 62) | (1 << 61);
+        const LEAF_4: u64 = 1 << 63;
+        const LEAF_5: u64 = (1 << 63) | (1 << 61);
+        const LEAF_6: u64 = (1 << 63) | (1 << 62);
+        const LEAF_7: u64 = (1 << 63) | (1 << 62) | (1 << 61);
+
+        let key_0 = Word::from([ZERO, ZERO, ZERO, Felt::new(LEAF_0)]);
+        let key_1 = Word::from([ZERO, ZERO, ZERO, Felt::new(LEAF_1)]);
+        let key_2 = Word::from([ZERO, ZERO, ZERO, Felt::new(LEAF_2)]);
+        let key_3 = Word::from([ZERO, ZERO, ZERO, Felt::new(LEAF_3)]);
+        let key_4 = Word::from([ZERO, ZERO, ZERO, Felt::new(LEAF_4)]);
+        let key_5 = Word::from([ZERO, ZERO, ZERO, Felt::new(LEAF_5)]);
+        let key_6 = Word::from([ZERO, ZERO, ZERO, Felt::new(LEAF_6)]);
+        let key_7 = Word::from([ZERO, ZERO, ZERO, Felt::new(LEAF_7)]);
+
+        // Create full SMT with keys 1 and 3 (key_3 makes node b non-empty)
+        let mut full = Smt::with_entries([(key_1, rand_value()), (key_3, rand_value())]).unwrap();
+
+        // Create partial SMT with ONLY the proof for key 1
+        let proof_1 = full.open(&key_1);
+        let mut partial = PartialSmt::from_proofs([proof_1]).unwrap();
+        assert_eq!(full.root(), partial.root());
+
+        // Key 1: CAN update (explicitly tracked via proof)
+        let new_value_1: Word = rand_value();
+        full.insert(key_1, new_value_1).unwrap();
+        partial.insert(key_1, new_value_1).unwrap();
+        assert_eq!(full.root(), partial.root());
+
+        // Key 0: CAN update (sibling of key 1, empty)
+        let value_0: Word = rand_value();
+        full.insert(key_0, value_0).unwrap();
+        partial.insert(key_0, value_0).unwrap();
+        assert_eq!(full.root(), partial.root());
+
+        // Key 4: CAN update (in empty subtree f)
+        let value_4: Word = rand_value();
+        full.insert(key_4, value_4).unwrap();
+        partial.insert(key_4, value_4).unwrap();
+        assert_eq!(full.root(), partial.root());
+
+        // Key 5: CAN update (in empty subtree f)
+        let value_5: Word = rand_value();
+        full.insert(key_5, value_5).unwrap();
+        partial.insert(key_5, value_5).unwrap();
+        assert_eq!(full.root(), partial.root());
```

Comment:
For key 5, 6, 7 it may be worth noting somehow that after updating key 4, the subtree is no longer empty. Just to make sure the test is not misunderstood easily.